### ISSUE TITLE
Wrong type argument `org-roam-node nil`

### ISSUE THREAD

**Issue by thiskappaisgrey** [emacs-citar/citar-org-roam#16]
This is a similar issue to #5 , but it's happening even with `citar-open`. The issue is that some of the id gets truncated when it is searched for, here's an example:
![image](https://user-images.githubusercontent.com/16870341/192127117-1487a622-ea91-4a14-b47f-a5c22269a040.png)
The top buffer is the file I'm trying to open with `citar-open`, and the bottom is the debug stack trace when I call `citar-open`. As you can see, the full id is: "58bcce67-c7f1-449e-b20a-4eba61fbd6d5" but when `citar-open-note` is called, some of the id gets truncated to:
"58bcce67-c7f1-449e-b20a-4e" which leads to the `org-roam-node nil` error.

**Comment by bdarcus** [comment]
I can't reproduce that, and am not really sure what could explain that.

BTW, @roshanshariff - why do we have that `citar-open-note` autoloaded interactive function AND `citar-open-notes`? We should remove one; no?

**Comment by bdarcus** [comment]
You can see what's going on in the code here:

https://github.com/emacs-citar/citar-org-roam/blob/29688b89ac3bf78405fa0dce7e17965aa8fe0dff/citar-org-roam.el#L97-L101

So it's just splitting that space-separated candidate string, and taking the `cadr`.

**Comment by thiskappaisgrey** [comment]
@bdarcus I think the bug is somewhere in `citar--get-resource-candidates`, since doing:
`(citar--get-resource-candidates '("wadlerMonadsFunctionalProgramming") :notes t :create-notes t)` yields the truncated id. The full id (which is the correct id, and will open the node) is "58bcce67-c7f1-449e-b20a-4eba61fbd6d5", but the function `citar--get-resource-candidates` returns an id that gets shortened/truncated to "4cc3c167-eace-491e-9175-0e" and I'm not really sure why either... I'm using `org-id-method uuid` to generate org-ids.

**Comment by bdarcus** [comment]
If I do the same thing, I get this, the only thing that looks odd to me is the trailing spaces:

```elisp
 (citar--get-resource-candidates '("toly2017") :notes t :create-notes t)
(org-roam-node
 #("toly2017 93f2c87d-716e-499c-b157-36f7d15e7334               " 0 9
   (citar--resource note)
   9 45
   (citar--resource note invisible t)
   45 60
   (citar--resource note)))
```

What do your candidates look like (`citar-org-roam--get-candidates)`?

**Comment by thiskappaisgrey** [comment]
`#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("wadlerMonadsFunctionalProgramming" (#("wadlerMonadsFunctionalProgramming 58bcce67-c7f1-449e-b20a-4e" 34 60 (invisible t)))))`

I think I understand the problem now, my citekey ("wadlerMonadsFunctionalProgramming") is too long! In the function:
`citar-org-roam--get-candidates`, there's this piece of code here:
```
(truncate-string-to-width
          (concat citekey " " (propertize nodeid 'invisible t)) 60 nil 32)
```
 the citekey and the nodeid gets `truncated`  *together*! In the example you sent, your citekey, _toly2017_ is pretty short, so it works just fine. I tried it on one of my notes with a short citekey, _ahrens2017_ and it works too.

**Comment by bdarcus** [comment]
> Is there a reason for truncating the string?

Now that I think of it ... we use that function really just for the display alignment; the intention is not that it truncates, but that it adds the padding.

That 60 number is really assuming nobody would ever have a citekey 60 characters long ;-)

So if not 60, then, what is a reasonably compromise?

**Comment by thiskappaisgrey** [comment]
Yeah, my keys were generated by Zotero, so people might accidentally have longer keys. Maybe you can store the uuid and the key in a 2 element list (I forgot what it's called..) then format it afterwards for display? As for the character limit, I'm fine with whatever you come up with, but I think it's a good idea to put in the readme that there is a character limit on the keys.

**Comment by bdarcus** [comment]
> Maybe you can store the uuid and the key in a 2 element list (I forgot what it's called..) then format it afterwards for display?

IIRC, we're limited to strings for the candidates.

**Comment by bdarcus** [comment]
OK, done. Let me know if that works for you?

**Comment by thiskappaisgrey** [comment]
Thanks!

**Comment by roshanshariff** [comment]
@bdarcus, `citar-open-note` opens a note given its ID, rather than citation key (as `citar-open-notes` does). We needed that function for non-interactive use. When called interactively it does completion based on the list of all notes provided by the note source, which could be useful in its own right.

**Comment by bdarcus** [comment]
> We needed that function for non-interactive use. 

Yes, I know this, but ...

> When called interactively it does completion based on the list of all notes provided by the note source, which could be useful in its own right.

... this part occurred to me is a bit confusing from `M-x`.

**Comment by roshanshariff** [comment]
> ... this part occurred to me is a bit confusing from M-x.

Hmm, I'm not sure how to avoid that. Perhaps by renaming one of those functions? If you feel it's warranted, you could remove the interactive specification of `citar-open-note`. Though, speaking for myself, I've found the interactive functionality of both commands useful.

**Comment by bdarcus** [comment]
I think a minimal change would just be to fix the docstring, so it's more clear when using marginalia, for example.

```
  "Open NOTE, which should be a string returned by `citar-get-notes'.
When called interactively, prompt for a note to open from a list
of all notes."
```

Maybe something like:

```
  "Open a single NOTE directly.
The note should be represented as a string returned by `citar-get-notes'.
When called interactively, prompt for a note to open from a list
of all notes."
```

**Comment by bdarcus** [comment]
Done; I think this is a little more clear.

![image](https://user-images.githubusercontent.com/1134/192168852-196f911e-4e93-4922-99e8-16d9c1d958d0.png)

**Comment by fuzzbomb** [comment]
I ran into this as well. Here are some notes about my own experience.

> That 60 number is really assuming nobody would ever have a citekey (+uuid) 60 characters long.
>
>But that was a bad assumption in retrospect.

Well an Org UUID is 36 characters long, so there was just  23 characters left for a cite key (and one space as separator).

With the new trunctation limit at 87 characters, there will be room for a cite key 50 characters long.
 
- Approximately 60% of my cite keys are longer than 23 characters. I ran into this problem a lot.
- Approximately 10% of my cite keys are longer than 40 characters.
- My longest cite key (currently) is 45 characters long.

I expect one day I will have a cite key which is longer than 50 characters, and _I will run into this problem again_.

> Yeah, my keys were generated by Zotero, so people might accidentally have longer keys

Long cite keys might be intentional, rather than accidental:
 
- Sometimes it's useful to mention all the authors in the cite key.
- Personally, I like cite keys which are meaningful when I read them, so I spend less time checking what work they refer to. Keys like `Stavrou2021b` aren't helpful for me, because I have lots of BibTex entries for Stavrou. So I use all (or enough) of the meaningful words from the work title.

Example: This shows how easy it is to arrive at a 37-character cite key. The key includes 3 surnames (one is double-barrelled) and 3 (English) title words.
 
```
@Online{WilliamsBaileyMaag2021DontBelieveType,
  author =       {Williams, Gareth and Bailey, David and Maag, Bruno},
  title =        {Don't Believe The Type!},
  year =         2021,
  ... more fields ...
}
```

Some ideas:

- Does the string have to be truncated at all?

- Does the string have to put the cite key first, and ID afterwards?  If the order were swapped, you could guarantee that a 36-character UUID wouldn't be truncated, no matter how long the cite key was. That should be long enough for timestamp-based IDs too. I haven't studied how Emacs text properties work yet, mind.

**Comment by bdarcus** [comment]
> Does the string have to be truncated at all?

Yes; per my point above, it's how alignment of the annotations is achieved.

> Does the string have to put the cite key first, and ID afterwards?

Am not sure, but suspect truncating the citekey would also be a problem elsewhere.

Another option may be to remove the annotation and put that content into the actual candidate string.

But handling alignment there would be a different type of problem (doable, without any downside; just more work).

**Comment by bdarcus** [comment]
Here's a completely different approach, that doesn't use an annotation function or `truncate-string-to-width`.

![image](https://user-images.githubusercontent.com/1134/192562833-cbd5a7d8-6fdd-44b6-8b99-63003998f8a5.png)

**Comment by bdarcus** [comment]
I have the last option implemented in the linked PR (#17).
